kbs/src/token/mod.rs

```diff
@@ -2,54 +2,86 @@
 // Licensed under the Apache License, Version 2.0, see LICENSE for details.
 // SPDX-License-Identifier: Apache-2.0
 
-use anyhow::*;
-use async_trait::async_trait;
+use jwk::JwkAttestationTokenVerifier;
+use kbs_types::TeePubKey;
 use serde::Deserialize;
-use std::sync::Arc;
+use serde_json::Value;
 use strum::EnumString;
-use tokio::sync::RwLock;
 
-mod coco;
+mod error;
 pub(crate) mod jwk;
+pub use error::*;
 
-#[async_trait]
-pub trait AttestationTokenVerifier {
-    /// Verify an signed attestation token.
-    /// Returns the custom claims JSON string of the token.
-    async fn verify(&self, token: String) -> Result<String>;
-}
+pub const TOKEN_TEE_PUBKEY_PATH_ITA: &str = "/attester_runtime_data/tee-pubkey";
+pub const TOKEN_TEE_PUBKEY_PATH_COCO: &str = "/customized_claims/runtime_data/tee-pubkey";
 
-#[derive(Deserialize, Default, Debug, Clone, EnumString)]
+#[derive(Deserialize, Default, Debug, Clone, EnumString, PartialEq)]
 pub enum AttestationTokenVerifierType {
     #[default]
     CoCo,
-    Jwk,
+
+    #[serde(rename = "ITA")]
+    Ita,
 }
 
-#[derive(Deserialize, Debug, Clone)]
+#[derive(Deserialize, Debug, Clone, PartialEq, Default)]
 pub struct AttestationTokenVerifierConfig {
     #[serde(default)]
-    pub attestation_token_type: AttestationTokenVerifierType,
+    pub r#type: AttestationTokenVerifierType,
 
-    /// Trusted Certificates file (PEM format) path (for "CoCo") or a valid Url
-    /// (file:// and https:// schemes accepted) pointing to a local JWKSet file
+    /// Trusted Certificates file (PEM format) paths use to verify Attestation
+    /// Token Signature.
+    #[serde(default)]
+    pub trusted_certs_paths: Vec<String>,
+
+    /// Urls (file:// and https:// schemes accepted) pointing to a local JWKSet file
     /// or to an OpenID configuration url giving a pointer to JWKSet certificates
     /// (for "Jwk") to verify Attestation Token Signature.
     #[serde(default)]
-    pub trusted_certs_paths: Vec<String>,
+    pub trusted_jwk_sets: Vec<String>,
+
+    /// Whether a JWK that directly comes from the JWT token is allowed to verify
+    /// the signature. This is insecure as it will not check the endorsement of
+    /// the JWK. If this option is set to false, the JWK will be looked up from
+    /// the key store configured during launching the KBS with kid field in the JWT,
+    /// or be checked against the configured trusted CA certs.
+    #[serde(default = "bool::default")]
+    pub insecure_key: bool,
```

Comment:
I tried hard not to break users then `Jwk` type was added. I don't see these changes absolutely necessary either. how would users know when to use trusted_certs_paths` or `trusted_jwk_sets`? IMO, for a proper token verifier unification, we don't have a "type" at all and users would only configure their trust sources.

Reply:
This is hard to do. Like I said in https://github.com/confidential-containers/trustee/issues/519#issuecomment-2398902746. Currently KBS will receive JWTs from either CoCoAS or ITA. I have no good idea to detect which concrete JWT is runtimely. So the design here is to configure the token type in the config file.

In future, I wonder if we can change the authentication way here. That means
1. Attestation module would be responsible to interactive with backend attestation services and get a `yes/no` result.
2. KBS will provision a certificate that certs the tee public key and an id inside.

In this way, the storage and other plugins will only need to check the endorsement of the client certificate and the  id. Then the policy engine (OPA) would do much easier job -- control API access due to the client id, rather than complex different attestation claims.